src/librustc_mir/transform/validate.rs

```diff
@@ -40,6 +40,40 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
     }
 }
 
+/// Check if src can be assigned into dest.
+/// This is not precise, it will accept some incorrect assignments.
+fn mir_assign_valid_types<'tcx>(tcx: TyCtxt<'tcx>, src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+    if src == dest {
+        // Equal types, all is good.
+        return true;
+    }
+
+    // Type-changing assignments can happen for (at least) two reasons:
+    // 1. `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
+    // 2. Subtyping is used. While all normal lifetimes are erased, higher-ranked types
+    //    with their late-bound lifetimes are still around and can lead to type differences.
+    // Normalize both of them away.
+    // FIXME: Share this code with `interpret/eval_context.rs`.
+    let normalize = |ty: Ty<'tcx>| {
+        ty.fold_with(&mut BottomUpFolder {
+            tcx,
+            // Normalize all references to immutable.
+            ty_op: |ty| match ty.kind {
+                ty::Ref(_, pointee, _) => tcx.mk_imm_ref(tcx.lifetimes.re_erased, pointee),
+                _ => ty,
+            },
```

Comment:
Aren't we supposed to not be doing this sort of thing anymore? If you remove this part, then you could use `normalizing_erasing_regions` for everything w/o late-bound lifetimes (which is cheap to check for, using flags), which should be a perf boost.

Reply:
I have no idea what we are supposed to do -- but I am also normalizing mutability here, so I don't know how to use `normalizing_erasing_regions` for that.

Reply:
Oh I might've been thinking of `Operand::Copy` of `&mut T` being removed.

So we still confuse `&mut T` and `&T` sometimes? Oh but isn't it only at the outermost level? So you could still remove it out of the recursion, and just have it before/after it (or even handle it in the comparison instead of normalization).

Reply:
It is probably only outermost.

But even then, IIRC `normalizing_erasing_regions` doesn't erase enough lifetimes -- we do need to erase higher-ranked lifetimes bound inside the type here as well due to subtyping.